--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -40,7 +40,8 @@
 #include "mtk_eth_reset.h"
 #include "mtk_hnat/nf_hnat_mtk.h"
 #endif
-
+struct net_device *ppd_dev;
+EXPORT_SYMBOL(ppd_dev);
 static int mtk_msg_level = -1;
 module_param_named(msg_level, mtk_msg_level, int, 0);
 MODULE_PARM_DESC(msg_level, "Message level (-1=defaults,0=none,...,16=all)");
@@ -2489,11 +2490,11 @@ static int mtk_poll_rx(struct napi_struc
 		}
 
 		if ( (mac == 4) || ((FIELD_GET(MTK_RXD5_PPE_CPU_REASON, trxd.rxd5)) == HIT_BIND_FORCE_TO_CPU)) 
-			
-		{
-			mac = 1;
+			{
 			sent_ppd = 1;
-		}
+			mac =0;
+			}
+
 
 		if (unlikely(mac < 0 || mac >= MTK_MAX_DEVS ||
 			     !eth->netdev[mac]))
@@ -2656,9 +2657,9 @@ static int mtk_poll_rx(struct napi_struc
 			hw_lro_flush_stats_update(ring->ring_no, &trxd);
 		}
 
-		if (sent_ppd){
+		if (sent_ppd && ppd_dev){
 			skb->protocol = eth_hdr(skb)->h_proto;
-			skb->dev = eth->netdev[1];
+			skb->dev = ppd_dev;
 		}
 		skb_record_rx_queue(skb, 0);
 		napi_gro_receive(napi, skb);
--- a/drivers/usb/serial/option.c
+++ b/drivers/usb/serial/option.c
@@ -48,6 +48,7 @@ static void option_release(struct usb_se
 static void option_instat_callback(struct urb *urb);
 
 /* Vendor and product IDs */
+#define TDTECH_VENDOR_ID			0x3466
 #define OPTION_VENDOR_ID			0x0AF0
 #define OPTION_PRODUCT_COLT			0x5000
 #define OPTION_PRODUCT_RICOLA			0x6000
@@ -639,6 +640,12 @@ static void option_instat_callback(struc
 
 
 static const struct usb_device_id option_ids[] = {
+	{ USB_VENDOR_AND_INTERFACE_INFO(TDTECH_VENDOR_ID, 0xff, 0x06, 0x10) },
+	{ USB_VENDOR_AND_INTERFACE_INFO(TDTECH_VENDOR_ID, 0xff, 0x06, 0x12) },
+	{ USB_VENDOR_AND_INTERFACE_INFO(TDTECH_VENDOR_ID, 0xff, 0x06, 0x13) },
+	{ USB_VENDOR_AND_INTERFACE_INFO(TDTECH_VENDOR_ID, 0xff, 0x06, 0x14) },
+	{ USB_VENDOR_AND_INTERFACE_INFO(TDTECH_VENDOR_ID, 0xff, 0x06, 0x1B) },
+	{ USB_VENDOR_AND_INTERFACE_INFO(TDTECH_VENDOR_ID, 0xff, 0x06, 0x1C) },
 	{ USB_DEVICE(OPTION_VENDOR_ID, OPTION_PRODUCT_COLT) },
 	{ USB_DEVICE(OPTION_VENDOR_ID, OPTION_PRODUCT_RICOLA) },
 	{ USB_DEVICE(OPTION_VENDOR_ID, OPTION_PRODUCT_RICOLA_LIGHT) },
